lib/dma.h

```diff
@@ -182,24 +211,33 @@ _dma_mark_dirty(const dma_controller_t *dma, const dma_memory_region_t *region,
     range_to_pages(sg->offset, sg->length, dma->dirty_pgsize,
                    &pgstart, &pgend);
 
-    index = bit_to_u8(pgstart);
-    end = bit_to_u8(pgend) + !!(bit_to_u8off(pgend));
-
-    for (i = index; i < end; i++) {
-        uint8_t bm = ~0;
-
-        /* Mask off any pages in the first u8 that aren't in the range. */
-        if (i == index && bit_to_u8off(pgstart) != 0) {
-            bm &= ~((1 << bit_to_u8off(pgstart)) - 1);
-        }
+    bm_sg_start = bit_to_u8(pgstart);
+    bm_sg_end = bit_to_u8(pgend) + !!(bit_to_u8off(pgend));
+
+    assert(bm_sg_start <= bm_sg_end);
+    
+    u8_set_all(&bm);
+    /* If only one byte of the bitmap has dirty pages, mark them and return. */
+    if (pgstart == pgend || bm_sg_start == bm_sg_end - 1) {
```

Comment:
what do we need the first pgstart == pgend check here?

Reply:
so if pgstart == pgend && pgstart % 8 == 0 then we have bm_sg_start == bm_sg_end. The condition could be rewritten:

if (bm_sg_start == bm_sg_end || bm_sg_start == bm_sg_end - 1)

I think that's a bit easier on the eyes so maybe I'll do that and add a comment.

Reply:
ok, but you can do that with a single less than /more than comparison